File: JavaScript/Data Structures/Linked Lists/SinglyLinkedList.js
class Node {
    constructor(val){
        this.val = val;
        this.next = null;
    }
}

class SinglyLinkedList {
    constructor(){
        this.head = null;
        this.tail = null;
        this.length = 0;
    }

    push(val){
        let node = new Node(val);
        if (this.length === 0) {
            this.head = node;
            this.tail = node;
        } else {
            this.tail.next = node;
            this.tail = node;
        }
        this.length +=1;
        return this;
    }

    pop(){
        if (this.length === 0){
            return;
        } 
        var current = this.head;
        var pre = current;
        while(current.next){
            pre = current;
            current = current.next;
        }
        this.tail = pre;
        pre.next = null;
        this.length -= 1;
        if (this.length === 0){
            this.head = null;
            this.tail = null;
        }
        return current;
    }

    shift(){
        if (this.length === 0) return;
        let node = this.head;
        this.head = node.next;
        this.length -= 1;
        if (this.length === 0){
            this.tail = null;
        }
        return node;
    }

    unshift(val){
        let node = new Node(val);
        if (this.length === 0){
            this.head = node;
            this.tail = node;
        } else {
            let currentHead = this.head;
            this.head = node;
            this.head.next = currentHead;
        }
        this.length +=1;
        return this;
    }

    get(idx){
        if (idx >= this.length || idx < 0){
            return;
        } 
        else if (idx === 0) {
            return this.head;
        }
        else {
            let current = this.head;
            for(let i = 1; i <= idx; i++){
                current = current.next;
            }
            return current;
        }
    }

    set(idx, val){
        var newNode = this.get(idx);
        if (newNode){
            newNode.val = val;
            return true;
        }
        return false;
    }

    insert(idx, val){
        if (idx > this.length || idx < 0){
            return false;
        } else if (idx === 0){
            let newNode = this.unshift(val);
            if (newNode){
                return true;
            } else {
                return false;
            }
        } else if (idx === this.length){
            let newNode = this.push(val);
            if (newNode){
                return true;
            } else {
                return false;
            }
        } else {
            let preNode = this.get(idx-1);
            let newNode = new Node(val);
            newNode.next = preNode.next;
            preNode.next = newNode;
            this.length += 1;
            return true;
        }
    }

    remove(idx){
        if (idx >= this.length || idx < 0){
            return;
        } else if (idx === 0){
            return this.shift();
        } else if (idx === this.length - 1){
            return this.pop();
        } else {
            let preNode = this.get(idx-1);
            let currentNode = this.get(idx);
            preNode.next = currentNode.next;
            currentNode.next = null;
            this.length -= 1;
            return currentNode;
        }
    }
}

let sll = new SinglyLinkedList()

sll.push(1)
sll.push(2)
sll.push(3)
sll.push(4)
console.log('--------------------')
console.log(sll.remove(2))
console.log(sll.remove(1))
console.log(sll.remove(1))
console.log(sll.remove(0))
//console.log(sll.get(0))
//console.log(sll.get(1))
//console.log(sll.get(2))
console.log(sll)

File: Python/Data Structures/LinkedLists/SinglyLinkedList.py
class Node:
    def __init__(self,val):
        self.val = val
        self.next = None

class SinglyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    def push(self,val):
        newNode = Node(val)
        if self.length == 0:
            self.head = newNode
            self.tail = newNode
        else:
            self.tail.next = newNode
            self.tail = newNode
        self.length += 1
        return self

    def pop(self):
        if self.length == 0:
            return None
        popped = self.tail
        if self.length == 1:
            self.head = None
            self.tail = None
        else:
            current = self.head
            previous = current
            while current.next:
                previous = current
                current = current.next
            self.tail = previous
            self.tail.next = None
        self.length -= 1
        return popped.val

    def shift(self):
        if self.length == 0:
            return None
        oldHead = self.head
        if self.length == 1:
            self.head = None
            self.tail = None
        else:
            self.head = oldHead.next
            oldHead.next = None
        self.length -= 1
        return oldHead.val

    def unshift(self, val):
        newNode = Node(val)
        if self.length == 0:
            self.head = newNode
            self.tail = newNode
        else:
            newNode.next = self.head
            self.head = newNode
        self.length += 1
        return self

    def get(self, idx):
        if self.length == 0 or idx < 0 or idx >= self.length:
            return None
        elif idx == 0:
            return self.head
        elif idx == self.length - 1:
            return self.tail
        else:
            current = self.head
            cnt = 1
            while cnt <= idx:
                cnt += 1
                current = current.next
            return current

    def set(self, idx, val):
        node = self.get(idx)
        if node:
            node.val = val
            return True
        else:
            return False

    def print(self):
        output = []
        if self.length == 0:
            return output
        else:
            current = self.head
            while current:
                output.append(current.val)
                current = current.next
        return output

sll = SinglyLinkedList()
print(sll.get(0))
print(sll.get(-10))
print(sll.get(10))
sll.unshift(1)
print(sll.print())
sll.unshift(2)
print(sll.print())
sll.unshift(3)
print(sll.print())
sll.unshift(4)
print(sll.print())
sll.push(5)
print(sll.print())
sll.push(6)
print(sll.print())
sll.set(0,10)
print(sll.print())
sll.set(1,11)
print(sll.print())
sll.set(2,12)
print(sll.print())
sll.set(3,13)
print(sll.print())
sll.set(4,14)
print(sll.print())
sll.set(5,15)
print(sll.print())
print(sll.set(6,15))
print(sll.print())
